### PR TITLE
Added clusterIP service type for provider server

### PR DESCRIPTION
**Pull request by rohan47**
On HCI cluster the hosted OCP cluster are on the same pod network as the Host OCP so a ClientIP service will be the better option than NodePort / LoadBalancer.

ProviderAPIServerServiceType needs to be updated in ocs-operator so that it takes ClusterIP as a value and creates the service

### CODE REVIEW

File: controllers/storagecluster/provider_server.go [hunk 1]
```diff
@@ -210,6 +210,8 @@ func (o *ocsProviderServer) createService(r *StorageClusterReconciler, instance
 
 		instance.Status.StorageProviderEndpoint = fmt.Sprintf("%s:%d", nodeAddresses[0], ocsProviderServiceNodePort)
 
+	} else {
+		instance.Status.StorageProviderEndpoint = fmt.Sprintf("%s:%d", actualService.Spec.ClusterIP, ocsProviderServicePort)
 	}
```
Comment (rewantsoni) [inline controllers/storagecluster/provider_server.go:215]: should use a switch case here instead of if-elseif-else?

**Review by rewantsoni — COMMENTED** [review]
Are you going to add unit tests in a separate PR/commit?

**Comment by Madhu-1** [comment]
> On HCI cluster the hosted OCP cluster are on the same pod network as the Host OCP so a ClientIP service will be the better option than NodePort / LoadBalancer.
> 
> ProviderAPIServerServiceType needs to be updated in ocs-operator so that it takes ClusterIP as a value and creates the service

Will this work for RDR for provider as well? or we are going to run new service between api-server for RDR?

**Comment by rohan47** [comment]
> > On HCI cluster the hosted OCP cluster are on the same pod network as the Host OCP so a ClientIP service will be the better option than NodePort / LoadBalancer.
> > ProviderAPIServerServiceType needs to be updated in ocs-operator so that it takes ClusterIP as a value and creates the service
> 
> Will this work for RDR for provider as well? or we are going to run new service between api-server for RDR?

There needs to be a nodePort service created so that provider services can reach each other. Also there needs to be a further discussion regarding this

File: controllers/storagecluster/provider_server.go (hunk 1 above)
Comment (rohan47) [reply controllers/storagecluster/provider_server.go:215]: will make the change

**Comment by leelavg** [comment]
> Will this work for RDR for provider as well? or we are going to run new service between api-server for RDR?
- @Madhu-1 do you mean, configuring both NodePort/LB and ClusterIP for apiServer or a new gRPC altogether?

File: controllers/storagecluster/provider_server.go [hunk 2]
```diff
@@ -147,8 +147,8 @@ func (o *ocsProviderServer) createDeployment(r *StorageClusterReconciler, instan
 func (o *ocsProviderServer) createService(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) (reconcile.Result, error) {
 
 	if instance.Spec.ProviderAPIServerServiceType != "" && instance.Spec.ProviderAPIServerServiceType != corev1.ServiceTypeNodePort &&
-		instance.Spec.ProviderAPIServerServiceType != corev1.ServiceTypeLoadBalancer {
-		err := fmt.Errorf("providerAPIServer only supports service of type %s and %s", corev1.ServiceTypeNodePort, corev1.ServiceTypeLoadBalancer)
+		instance.Spec.ProviderAPIServerServiceType != corev1.ServiceTypeLoadBalancer && instance.Spec.ProviderAPIServerServiceType != corev1.ServiceTypeClusterIP {
```
Comment (leelavg) [inline controllers/storagecluster/provider_server.go:150]: - this line is getting busy, maybe simplify w/ `switch/case` 🤔 
``` go
if instance.Spec.ProviderAPIServerServiceType != "" {
  switch instance.Spec.ProviderAPIServerServiceType {
    case corev1.ServiceTypeClusterIP, corev1.ServiceTypeNodePort, corev1.ServiceTypeLoadBalancer:
    default:
      // log and return error
  }
}

**Comment by Madhu-1** [comment]
> > Will this work for RDR for provider as well? or we are going to run new service between api-server for RDR?
> 
> * @Madhu-1 do you mean, configuring both NodePort/LB and ClusterIP for apiServer or a new gRPC altogether?

i don't know the plan around RDR yet but i was curious to know more about it what is the use of clusterIP server and are we going to have two services for the provider server? With the current code it looks like only one service with clusterIP is created but in RDR i don't think two clusters will be able to talk over clusterIP (but I could be wrong as i don't know much about architecture of it yet)

File: controllers/storagecluster/provider_server.go [hunk 3]
```diff
@@ -384,7 +394,7 @@ func GetProviderAPIServerService(instance *ocsv1.StorageCluster) *corev1.Service
 			},
 			Ports: []corev1.ServicePort{
 				{
-					NodePort:   ocsProviderServiceNodePort,
+					NodePort:   nodePort,
```
Comment (leelavg) [inline controllers/storagecluster/provider_server.go:397]: - this is a bit unusal, either add a comment why default case (clusterip) is inverted or refactor like ...
``` go
servicePort := corev1.ServiePort{Port: ocsProviderServicePort, TargetPort: intstr.FromString("ocs-provider")}
if instance.Spec.ProviderAPIServerServiceType == corev1.ServiceTypeNodePort {
  servicePort.NodePort = ocsProviderServiceNodePort
}
// ...
Ports: []corev1.ServicePort{servicePort}
```
- I feel it's better not to set default case explicitly

File: controllers/storagecluster/provider_server.go (hunk 2 above)
Comment (rohan47) [reply controllers/storagecluster/provider_server.go]: Done

File: controllers/storagecluster/provider_server.go (hunk 1 above)
Comment (rohan47) [reply controllers/storagecluster/provider_server.go]: Done

**Comment by rohan47** [comment]
Testing the code changes

File: controllers/storagecluster/provider_server.go [hunk 4]
```diff
@@ -196,7 +201,11 @@ func (o *ocsProviderServer) createService(r *StorageClusterReconciler, instance
 		}
 
 		instance.Status.StorageProviderEndpoint = fmt.Sprintf("%s:%d", endpoint, ocsProviderServicePort)
-	} else {
+
+	case corev1.ServiceTypeClusterIP:
+		instance.Status.StorageProviderEndpoint = fmt.Sprintf("%s:%d", actualService.Spec.ClusterIP, ocsProviderServicePort)
+
+	default:
```
Comment (iamniting) [inline controllers/storagecluster/provider_server.go:208]: Can you pls add a comment here that the default value is Node port that's why we are adding it in default case and someone not change this.

**Review by iamniting — CHANGES_REQUESTED** [review]

File: controllers/storagecluster/provider_server.go (hunk 4 above)
Comment (rohan47) [reply controllers/storagecluster/provider_server.go:209]: Done

File: controllers/storagecluster/provider_server.go (hunk 3 above)
Comment (rohan47) [reply controllers/storagecluster/provider_server.go]: Done

**Comment by leelavg** [comment]
@Madhu-1, seems I missed your reply

> but i was curious to know more about it what is the use of clusterIP server
- this is more tailored to HCI wherein, Kubevirt instances are also on pod network (new to me) and so client operator can reach provider within the cluster

> are we going to have two services for the provider server?
- ya, that is TBD and at the same time, addition is easier than deletion and so I believe going with one service is ok for now

> but in RDR i don't think two clusters will be able to talk over clusterIP (but I could be wrong ...
- I don't think you are wrong, I'd defer to @rewantsoni to come with arguments or opinions that has to do anything with RDR for HCI.

File: controllers/storagecluster/provider_server.go [hunk 5]
```diff
@@ -196,7 +201,12 @@ func (o *ocsProviderServer) createService(r *StorageClusterReconciler, instance
 		}
 
 		instance.Status.StorageProviderEndpoint = fmt.Sprintf("%s:%d", endpoint, ocsProviderServicePort)
-	} else {
+
+	case corev1.ServiceTypeClusterIP:
+		instance.Status.StorageProviderEndpoint = fmt.Sprintf("%s:%d", actualService.Spec.ClusterIP, ocsProviderServicePort)
+
+	// Nodeport is the default ServiceType for the provider server
+	default:
```
Comment (iamniting) [inline controllers/storagecluster/provider_server.go:209]:
```suggestion
	default: // Nodeport is the default ServiceType for the provider server
```

File: controllers/storagecluster/provider_server.go [hunk 6]
```diff
@@ -146,11 +146,15 @@ func (o *ocsProviderServer) createDeployment(r *StorageClusterReconciler, instan
...
-	if instance.Spec.ProviderAPIServerServiceType != "" && instance.Spec.ProviderAPIServerServiceType != corev1.ServiceTypeNodePort &&
-		instance.Spec.ProviderAPIServerServiceType != corev1.ServiceTypeLoadBalancer {
-		err := fmt.Errorf("providerAPIServer only supports service of type %s and %s", corev1.ServiceTypeNodePort, corev1.ServiceTypeLoadBalancer)
-		r.Log.Error(err, "Failed to create/update service, Requested ServiceType is", "ServiceType", instance.Spec.ProviderAPIServerServiceType)
-		return reconcile.Result{}, err
+	if instance.Spec.ProviderAPIServerServiceType != "" {
+		switch instance.Spec.ProviderAPIServerServiceType {
+		case corev1.ServiceTypeClusterIP, corev1.ServiceTypeLoadBalancer, corev1.ServiceTypeNodePort:
+		default:
+			err := fmt.Errorf("providerAPIServer only supports service of type %s, %s and %s", corev1.ServiceTypeNodePort, corev1.ServiceTypeLoadBalancer, corev1.ServiceTypeClusterIP)
```
Comment (iamniting) [inline controllers/storagecluster/provider_server.go:153]: Can you pls divide it in 2 lines for better readability?

```
err := fmt.Errorf("providerAPIServer only supports service of type %s, %s and %s", 
    corev1.ServiceTypeNodePort, corev1.ServiceTypeLoadBalancer, corev1.ServiceTypeClusterIP)
```

**Review by iamniting — APPROVED** [review]

**Comment by iamniting** [comment]
@rohan47 Can you pls make above changes mistakenly approved it.

File: controllers/storagecluster/provider_server.go (hunk 5 above)
Comment (rohan47) [reply controllers/storagecluster/provider_server.go]: done

File: controllers/storagecluster/provider_server.go (hunk 6 above)
Comment (rohan47) [reply controllers/storagecluster/provider_server.go]: done

**Review by iamniting — APPROVED** [review]
changes look good to me. holding it for others' review.

/hold

**Comment by rohan47** [comment]
> @Madhu-1, seems I missed your reply
> 
> > but i was curious to know more about it what is the use of clusterIP server
> 
>     * this is more tailored to HCI wherein, Kubevirt instances are also on pod network (new to me) and so client operator can reach provider within the cluster
> 
> 
> > are we going to have two services for the provider server?
we can change the service type in the storagecluster and the change will reflect
> 
>     * ya, that is TBD and at the same time, addition is easier than deletion and so I believe going with one service is ok for now
> 
> 
> > but in RDR i don't think two clusters will be able to talk over clusterIP (but I could be wrong ...
> 
Two clusters cant talk over clusterIP
>     * I don't think you are wrong, I'd defer to @rewantsoni to come with arguments or opinions that has to do anything with RDR for HCI.
We can change the service type when required for RDR or create a temporary node port service. What do you say @rewantsoni

**Pull request merged**